# Implementing a Naive Bayes QA system

In [1]:
import pandas as pd
import os
import spacy
import numpy as np
from sklearn.model_selection import train_test_split

Data = df1.csv <br>
Format = [question, answer, yes/no]

### Importing data - Quartz

In [2]:
#myData = pd.read_csv('df1.csv')
myData = pd.read_csv(open('df1.csv'),sep='|')

In [3]:
myData = myData.drop('Unnamed: 0', axis= 1)

In [4]:
X = myData[['question.stem','text']]
y = myData[['y']]

In [5]:
myData.head()

question.stem        text  y
0  John's town used to have lots of water, back w...      scarce  1
1  John's town used to have lots of water, back w...   plentiful  0
2  Electrons further away from a nucleus have ___...      higher  1
3  Electrons further away from a nucleus have ___...       lower  0
4  Milo threw both a basketball and a baseball th...  basketball  1

### Train Test split 

In [6]:
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.33, random_state=42)

### Concat

In [7]:
train = pd.concat([X_train,y_train], axis = 1)

In [8]:
test = pd.concat([X_test,y_test], axis = 1)

### Filter

In [9]:
train_1 = train[train['y']==1]

In [10]:
train_0 = train[train['y']==0]
test_1 = test[test['y']==1]
test_0 = test[test['y']==0]


### Convert each question into txt file 

#### Change working directory

In [11]:
os.chdir('/Users/suguthansekar/Library/Mobile Documents/com~apple~CloudDocs/Mine/MS/Course work/Semester 3/Machine learning lab/Dataset/ktrain dataset format/datadir/test/class1/')

#### for loop

In [12]:
for line in range(len(test_1)):
    f = open(str(line)+".txt", "w+")
    #f.write("This is a test ")
    myText = test_1.iloc[line,0] + " " + test_1.iloc[line,1]
    f.write(test_1.iloc[line,0] + " " + test_1.iloc[line,1])
    f.close()

#### Write a txt file

In [13]:
train_1.iloc[3,0] + " " + train_1.iloc[3,1]

'When you send more power to a magnet it becomes stronger'

In [14]:
type(train_1.iloc[line,0] + " " + train_1.iloc[line,1])

str

### BoW creation

In [15]:
# Import CountVectorizer
from sklearn.feature_extraction.text import CountVectorizer
from sklearn.naive_bayes import MultinomialNB

In [16]:
# Create a CountVectorizer object
vectorizer = CountVectorizer(lowercase=True, stop_words='english')

# Fit and transform X_train
X_train_bow = vectorizer.fit_transform(X_train['question.stem'])

# Transform X_test
X_test_bow = vectorizer.transform(X_test['question.stem'])

# Print shape of X_train_bow and X_test_bow
print(X_train_bow.shape)
print(X_test_bow.shape)

(3612, 2799)
(1780, 2799)


### Naive Bayes Classifier - With BoW

In [17]:
# Create a MultinomialNB object
clf = MultinomialNB()


# Fit the classifier
clf.fit(X_train_bow, y_train)

# Measure the accuracy
accuracy = clf.score(X_test_bow, y_test)
print("The accuracy of the classifier on the test set is %.3f" % accuracy)

# Predict the sentiment of a negative review
review = "Higher SPF leads to higher joy. If there is lower SPF, there is lower joy"
prediction = clf.predict(vectorizer.transform([review]))[0]
print("The classification provided by the classifier is %i" % (prediction))

The accuracy of the classifier on the test set is 0.273
The classification provided by the classifier is 1


/Users/suguthansekar/anaconda3/envs/myPython/lib/python3.6/site-packages/sklearn/utils/validation.py:724: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


### Result

Naive Bayes classifier performs at 27.3 % accuracy

### N-gram model 

In [ ]:
# Generate n-grams upto n=1
vectorizer_ng1 = CountVectorizer(ngram_range=(1,1))
ng1 = vectorizer_ng1.fit_transform(corpus)

In [39]:
# Generate n-grams upto n=2
vectorizer_ng2 = CountVectorizer(ngram_range=(1,2))
ng2 = vectorizer_ng2.fit_transform(corpus)

NameError: name 'corpus' is not defined

## Trying out different N-grams

### NG-2 model 

In [51]:
# Generate n-grams upto n=3
vectorizer_ng2 = CountVectorizer(ngram_range=(1, 2))

In [52]:
# Fit and transform X_train
X_train_bow = vectorizer_ng2.fit_transform(X_train['question.stem'])

# Transform X_test
X_test_bow = vectorizer_ng2.transform(X_test['question.stem'])

# Print shape of X_train_bow and X_test_bow
print(X_train_bow.shape)
print(X_test_bow.shape)

(3612, 18276)
(1780, 18276)


### NG-3 model

In [54]:
# Generate n-grams upto n=3
vectorizer_ng3 = CountVectorizer(ngram_range=(1, 3))

In [55]:
# Fit and transform X_train
X_train_bow = vectorizer_ng3.fit_transform(X_train['question.stem'])

# Transform X_test
X_test_bow = vectorizer_ng3.transform(X_test['question.stem'])

# Print shape of X_train_bow and X_test_bow
print(X_train_bow.shape)
print(X_test_bow.shape)

(3612, 41397)
(1780, 41397)


### NG-8 model 

In [57]:
# Generate n-grams upto n=3
vectorizer_ng8 = CountVectorizer(ngram_range=(1, 8))

In [58]:
# Fit and transform X_train
X_train_bow = vectorizer_ng8.fit_transform(X_train['question.stem'])

# Transform X_test
X_test_bow = vectorizer_ng8.transform(X_test['question.stem'])

# Print shape of X_train_bow and X_test_bow
print(X_train_bow.shape)
print(X_test_bow.shape)

(3612, 165838)
(1780, 165838)


### Naive Bayes Classifier - With N-Grams

In [59]:
# Create a MultinomialNB object
clf = MultinomialNB()


# Fit the classifier
clf.fit(X_train_bow, y_train)

# Measure the accuracy
accuracy = clf.score(X_test_bow, y_test)
print("The accuracy of the classifier on the test set is %.3f" % accuracy)

# Predict the sentiment of a negative review
review = "Higher SPF leads to higher joy. If there is lower SPF, there is lower joy"
#prediction = clf.predict(vectorizer.transform([review]))[0]
#print("The classification provided by the classifier is %i" % (prediction))

The accuracy of the classifier on the test set is 0.168


/Users/suguthansekar/anaconda3/envs/myPython/lib/python3.6/site-packages/sklearn/utils/validation.py:724: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
